service/telemetry.go

```diff
@@ -214,11 +213,16 @@ func (tel *telemetryInitializer) shutdown() error {
 	view.Unregister(tel.views...)
 
 	var errs error
+	errs = multierr.Append(errs, tel.tp.Shutdown(context.Background()))
+	errs = multierr.Append(errs, tel.mp.Shutdown(context.Background()))
+
 	for _, server := range tel.servers {
 		if server != nil {
 			errs = multierr.Append(errs, server.Close())
 		}
 	}
+	tel.serverWG.Wait()
```

Comment:
Hey @michalpristas, good question! From my understanding, I don't believe this will ever block on shutdown.

This `Wait` call will only be hit once [`Close`](https://pkg.go.dev/net/http#Server.Close) has been called on every relevant server. Relevant servers in this case (`tel.servers`) are all started in the goroutine that you've referenced in `InitPrometheusServer`, so every started server increments the wg's counter. 

I'll paste the concerning `asyncErrorChannel` reference in question:
```
		if serveErr := server.ListenAndServe(); serveErr != nil && !errors.Is(serveErr, http.ErrServerClosed) {
			asyncErrorChannel <- serveErr
		}
```

The [`ListenAndServe` method](https://pkg.go.dev/net/http#Server.ListenAndServe) blocks until an error is hit, or `Shutdown`/`Close` is called. In the case of an error being returned by the server, a message is sent to the `asyncErrorChannel` channel and the goroutine will exit calling `wg.Done`. For the case of `Shutdown` or `Close` being called, `ListenAndServe` documentation states that the error is returned **after** the server is closed. No errors can be sent from the server once the server is closed, they'd all be sent beforehand. From these cases, I don't believe we have to worry about `Wait` blocking in `shutdown`.

Side note, this wait group was actually added because of the exact concern you're raising, if I'm following right. The test [`TestServiceTelemetryCleanupOnError`](https://github.com/open-telemetry/opentelemetry-collector/blob/8131d31601b886ad90b2cc8ae2e83ea96a9e2ca6/service/service_test.go#L236) was leaking a goroutine intermittently. The test would actually attempt to start a server on the same port twice in a row, and if it hit it fast enough, the second bind request would be made before `Close` was complete. This would cause an error to be sent to the `asyncErrorChannel` after the test was complete, causing the situation you've referenced.

If `shutdown` is called, it should be able to handle the running goroutines for any situation. Let me know if I misunderstood though.

Reply:
the problem i'm seeing is very sensitive to timing and we may not see it most of the time.
but i still think that replacing
```
asyncErrorChannel <- serveErr
```

with 
```
select{
case asyncErrorChannel <- serveErr:
default:
}
```

would not be that intrusive and could prevent leaks/locks on shutdown

Reply:
My apologies, but I'm not seeing how this will prevent a leak or stop the `tel.serverWG.Wait()` call from blocking. Can you share more detail on how the timing of the channel send operation `asyncErrorChannel <- serveErr` could cause a problem here? My understanding is that within this function there's no concurrency. The operation sending `serveErr` to `asyncErrorChannel` blocks until it's complete. If `asyncErrorChannel` is already closed it will panic. Is there some possible situation where the send operation will block forever, and that would cause the blocking `Wait` call you've referenced?

I agree your proposed `select` statement isn't a huge impacting call, but I want to avoid introducing timing delays as a solution as much as possible.

Reply:
I've done some more research and seen that this is a pretty technique in go, so I've added a `1s` timeout. I'd still be interested to hear though how it could block in this situation.